### PR TITLE
Fixes #31370 - multi-host stub template snapshots

### PR DESCRIPTION
**Pull request by lzap**
Our snapshot template render test and generator only supports one host stub (ipv4 single NIC redhat OS). In order to test more ERB codepaths, we need a way to create snapshots against multiple host stubs (e.g. IPv6, dualstack, static IPv4 or IPv6, debian family). In order to do that, template metadata can be extended to define list of host stubs to create and test snapshots against.

Snapshot files are now named `ProvisioningTemplate/PXEGrub/Kickstart_default_PXEGrub.host4and6.snap.txt ` where `host4and6` is the stub name, I implemented the following for starters:

* host - the original simple host
* host6 - IPv6 provisioning NIC
* host4and6 - dualstack provisioning NIC
* host4static - single stack IPv4 with static network credentials
* host6static - single stack IPv4 with static network credentials

All stubs are Redhat family OS, we can add more families later (e.g. test debian templates against debian OS). I want this patch to be small, this can be implemented later. I took a chance to remove spaces from snapshot files, they change quite often and `git add` command is now easier.

Finally, there was a suggestion to support dualstack provisioing and while Foreman design is not ideal, after I implemented those snapshot changes I see that our templates do render fine so I am proposing to change the explicit error to just to advertise this via a kernel command line option and leaving the rest up to the users.

https://github.com/theforeman/foreman/pull/8145

### CODE REVIEW

File: app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb
Comment (lzap) [inline app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb:57]: These changes are coming from https://github.com/theforeman/foreman/pull/8145 and while I'd prefer explicit error, I think it's safer now to allow this since we have a better test coverage.

File: lib/foreman/renderer/source/snapshot.rb
Comment (lzap) [inline lib/foreman/renderer/source/snapshot.rb:21]: This change removes spaces from snapshot text files converting them to underscore.

File: app/services/foreman/template_snapshot_service.rb
Comment (ekohl) [inline app/services/foreman/template_snapshot_service.rb]: This part is mostly the same between all of these. Can we extract that to a helper?

File: app/views/unattended/provisioning_templates/PXEGrub/kickstart_default_pxegrub.erb
Comment (ekohl) [inline app/views/unattended/provisioning_templates/PXEGrub/kickstart_default_pxegrub.erb:9]: Can we phrase this in a way that's also tells the user something? Perhaps `networking`?

File: app/views/unattended/provisioning_templates/PXEGrub/kickstart_default_pxegrub.erb
Comment (ekohl) [inline app/views/unattended/provisioning_templates/PXEGrub/kickstart_default_pxegrub.erb]: Thoughts about renaming this to `host4` or even `host4dhcp`?

File: app/services/foreman/template_snapshot_service.rb
Comment (lzap) [reply app/services/foreman/template_snapshot_service.rb]: I prefer verbose version for tests or test stubs. Helpers make code less unreadable in this case.

File: app/views/unattended/provisioning_templates/PXEGrub/kickstart_default_pxegrub.erb
Comment (lzap) [reply app/views/unattended/provisioning_templates/PXEGrub/kickstart_default_pxegrub.erb:9]: Well stubs can change anything, not just networking. Better ideas?

File: app/views/unattended/provisioning_templates/PXEGrub/kickstart_default_pxegrub.erb
Comment (lzap) [reply app/views/unattended/provisioning_templates/PXEGrub/kickstart_default_pxegrub.erb]: Yeah I was thinking the same, I will do this.

File: app/services/foreman/template_snapshot_service.rb
Comment (lzap) [reply app/services/foreman/template_snapshot_service.rb]: Also note I have already extracted some things into `define_host_params` but other lines do not make much sense to me. It's hard to change things than and as I said less readable.

Comment (ekohl) [reply app/services/foreman/template_snapshot_service.rb]: Perhaps it could be a [trait in FactoryBot](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#traits)? So you don't define the arch, medium, ptable and OS every time. That already saves a lot of duplication.

Comment (lzap) [reply app/services/foreman/template_snapshot_service.rb]: That's a good idea.

**Comment by lzap** [comment]
One more thing this patch could solve is using these "test_on" tags to solve an ugly solution that was introduced some time ago: https://github.com/theforeman/foreman/pull/6951

My thinking is introducing test flag like `with_foreman_ansible` and that would cause the stub to actually return true on `plugin_present?`. This would give us full control on what to actually render during test.

**Comment by lzap** [comment]
I was fighting with tests a bit, now I am going to incorporate the comments.

**Comment by lzap** [comment]
Tests are green! I actually have one high-priority workaround we need to backport pretty soon which would benefit from this PR merged - I could add a specific test for EL 8.3. Thanks :-)

**Comment by ronnybremer** [comment]
Thanks Lukas for all the work you put in.

File: test/factories/host_related.rb
Comment (ezr-ondrej) [inline test/factories/host_related.rb:339]: can't we just have these as default values in the `TemplateSnapshotService`?
It doesn't really make much sense to have factory, that is only to define some fixed values for one use case and will not be used anywhere else :thinking:

File: app/views/unattended/provisioning_templates/PXEGrub/kickstart_default_pxegrub.erb
Comment (ezr-ondrej) [reply app/views/unattended/provisioning_templates/PXEGrub/kickstart_default_pxegrub.erb:9]: Maybe `supported_on`? and make it `ipv4dhcp`, `ipv4and6dhcp`? :thinking:

**Review by ezr-ondrej — CHANGES_REQUESTED** [review]
Thanks for all the hard work @lzap looks great...
Only one complaint in general: I don't like to pollute factories with non-generally useful edge cases, that make it hard to use factories, could we instead set the values in the Snapshot service directly? We can have some method with shared defaults for the host factory, right?

**Comment by lzap** [comment]
> Only one complaint in general: I don't like to pollute factories with non-generally useful edge cases

Eh, well me neither. All your comments are valid, however this was asked by @ekohl above. On the other hand, factories directory the right place and I tried to create as descriptive names as possible. So if you ask me, I lean towards keeping them in factories, I can move them back if you want. Whoever is gonna merge my patch I am gonna please :-)

> Maybe supported_on? and make it ipv4dhcp, ipv4and6dhcp

Support is a weird term in this context. I think I like "test_on" better.

Let me know if you insist on moving those fixtures back to the original place. I am rebasing in the meantime.

Note for myself: There are two other PRs that needs to be rebased on top of this.

**Comment by ezr-ondrej** [comment]
> On the other hand, factories directory the right place and I tried to create as descriptive names as possible.

I don't agree factories are the right place, but I'm ok with that, if you could just put them in one separate file eg `template_snapshot_related`?

`test_on` is fine by me. It's not perfect, but it works :shrug:

File: lib/tasks/snapshots.rake
Comment (ekohl) [inline lib/tasks/snapshots.rake]: Isn't this the same as
```suggestion
            File.write(snapshot_path, snapshot)
```

File: test/factories/domain.rb
Comment (ekohl) [inline test/factories/domain.rb]: Should we use `example.com`? or `snap.example.com`

File: test/factories/host_related.rb
Comment (ekohl) [inline test/factories/host_related.rb:327]: This looks like the wrong way to build associations if I read
https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#associations
```suggestion
      operatingsystem factory: :for_snapshots_centos_7_0
```

File: test/factories/host_related.rb
Comment (ekohl) [inline test/factories/host_related.rb:323]: Why a separate factory? Shouldn't it be a [trait](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#defining-traits)?

**Comment by lzap** [comment]
> I don't agree factories are the right place, but I'm ok with that, if you could just put them in one separate file eg `template_snapshot_related`?

Doesn't this nullify the purpose of a common place? To me this sounds like the worst solution, I'd rather have it in the test itself or in the host factory file.

Honestly, I think there is little to zero chance factories will ever get reused. But it was a huge pain to get it working. If we agree here, I am willing to build those hosts in the test itself like it was originally filed.

File: test/factories/domain.rb
Comment (lzap) [reply test/factories/domain.rb]: I wanted to differentiate this from exampleN.com but sure. Oh the subdomain is a good idea.

File: test/factories/host_related.rb
Comment (lzap) [reply test/factories/host_related.rb:327]: Well, that does not work. I know very little about FactoryBot and it took me a while to get this working honestly. Note I am not creating those records, I am building them (not saving them in the db).

I am leaving this as-is feel free to show me a working example.

File: test/factories/host_related.rb
Comment (lzap) [reply test/factories/host_related.rb:323]: That does not work either, I have no clue why but it looks like for traits some orchestration is being triggered.

Edit: I think `after(:build)` causes troubles. I had to create a new factory for this.

**Comment by lzap** [comment]
I rebased.

**Comment by ezr-ondrej** [comment]
> Honestly, I think there is little to zero chance factories will ever get reused. But it was a huge pain to get it working. If we agree here, I am willing to build those hosts in the test itself like it was originally filed.

I'd lean to having defaults defined in tests as having single purpose factories is quite an antipatern, but it doesn't hurt much eighter, I'm most worried of setting a bad example for the future, but I'm ok to merge anyway if we insist.

**Comment by ekohl** [comment]
Since I originally suggested it, let me explain. I suggested to move common patterns to factories. For example, some NIC or OS flavor is used multiple times. That is a good factory/trait. Things that are only used once isn't. So there is some compromise somewhere in between.

**Review by ezr-ondrej — COMMENTED** [review]
@lzap can you rebase and just state how you've decide make the data?
I'll merge even if you say you want to keep it in factories :)

**Comment by ezr-ondrej** [comment]
@lzap? I believe there is not much left to discuss and whatever you choose to implement I'm fine with that, it increases our test coverage, so I'm all up for that even if it's not implemented in the perfect way.

**Comment by lzap** [comment]
Weird, while I am rebasing this patch and regenerating snapshots, I am getting: The snippet 'puppet.conf' threw an error: undefined method '#present?' for Safemode::Jail (Environment) puppet.conf:102:in `bind'

https://github.com/theforeman/foreman/commit/c1104d0046d7d5d53bac7753b9be0b35361d9ef1#diff-3382905c2ea4ecd0ebd292ac3f720b45e89972a2d3f24a6f1d8f1c6586ffc419R21

We do have many `present?` calls, I don't get it. This method should be allowed for years now:

```
commit 0560dacaed5232df6f5ecbc848da608236ab3992
Author: Stephen Benjamin <stbenjam@xxx.com>
Date:   Fri Jan 8 10:51:02 2016 -0500

    Allow any?, blank?, present? in core jails
```

**Comment by lzap** [comment]
What is weird is it only fails for the `puppet.conf` snippet, we have many other occurrences and these work fine.

**Comment by lzap** [comment]
I think I have found the problem, method `present?` is allowed in Ruby core Object in safemode, however it is also redefined in ActiveRecord::Core. And in rendering tests whole Environment instance is stubbed instead of String.

I think it is safe to allow `present?` method from ActiveRecord, it returns always true anyway. I am doing this in the ActiveRecord base class in Foreman because I think this is where it belongs - we do have an apipie documentation group that documents allowing of id and name however this is actually not done so the documentation is misleading.

**Comment by lzap** [comment]
There is one more issue I need to solve, unattended_url setting is set in fixtures to `foreman.some.host.fqdn` but in `settings.yaml.test` to `foreman.example.com` and this creates problem in snapshot tests. While I could override the value, I think the cleaner approach would be to simple fix everything to the preferred "example.com" domain which is often recommended by @ekohl - so let me do that in this PR, tests will not pass until this is done.

Note for me - the commit before this change is eae4dc3626d10fffa081138739ca646d7800668a

File: test/fixtures/settings.yml
Comment (ezr-ondrej) [inline test/fixtures/settings.yml]: :heart:

**Comment by ezr-ondrej** [comment]
Though I agree changing the domain is a good way to go, I'd prefer it to be at least in separate commit, maybe even PR for better visibility and cleaner history. Also this should align with the Setting changes, where I'll need to touch these as well as fixtures won't be viable to change the default value anymore.

**Comment by lzap** [comment]
> Though I agree changing the domain is a good way to go, I'd prefer it to be at least in separate commit, maybe even PR for better visibility and cleaner history. Also this should align with the Setting changes, where I'll need to touch these as well as fixtures won't be viable to change the default value anymore.

Do you insist on this? Thing is that the new factories already use domain `example.com` and I thought it is a good time to actually change everything to example.com. I'd probably need to change everything to that weird some.host.fqdn (including the new factories) only after I would change it back in another PR.

**Comment by lzap** [comment]
I fixed the template change which was caused by having Ansible plugin. There is a PR to fix that scheduled to be merged after this one is merged :-)

**Comment by ezr-ondrej** [comment]
[test katello]

**Review by ezr-ondrej — DISMISSED** [review]
I believe this is worth merging even though we would cause trouble to stabilization, and it is not much chance to cause trouble as it touches mostly just tests.
But OTH, nothing is really forcing us to get it in 2.5 IIUC.
@tbrisker ? :)

**Comment by lzap** [comment]
I would love to get this in, I spent quite time on this and it vastly improves test coverage of templates.

I am rebasing due to conflict caused by the other PR.

**Comment by tbrisker** [comment]
@ezr-ondrej @lzap i'm fine with merging this during stabilization unless there is risk this could cause plugin tests to fail - in which case, let's wait a week and give plugin maintainers time to make sure their plugins are working against 2.5

**Comment by lzap** [comment]
Marking my calendar then, week it is, thanks :-)

**Comment by lzap** [comment]
I think I got it incorrectly, I do not thing this should introduce any problems to plugin authors. It just adds new contexts to the template renderer tests. There should not be any risk I am aware of.

File: app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb
Comment (tbrisker) [inline app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb:35]: This and the other change to this file do not seem like they are related to this issue and probably deserve their own separate PR

File: app/models/application_record.rb
Comment (tbrisker) [inline app/models/application_record.rb:15]: Also not related change, looks like `:name` comes from safemode itself: https://github.com/svenfuchs/safemode/blob/master/lib/safemode/core_jails.rb#L127
and i'm not sure about `:id` but it seems that it is already allowed from somewhere?

File: app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb
Comment (lzap) [reply app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb:35]: Ah this is a leftover from https://github.com/theforeman/foreman/pull/8447 that should not be here for sure.

File: app/models/application_record.rb
Comment (lzap) [reply app/models/application_record.rb:15]: I could not find where exactly `id` is allowed, the point of this hunk is that this class has documentation that says these are all allowed (I also added present? on top of that which is safe to add). Problem is, it was unclear where it gets added. So I wanted to explicitly name all three here, Safemode seems to walk up the class hierarchy when finding allowed methods and it is okay to "override" them.

Do you prefer not to mention `name` since it is allowed for all Ruby objects? For the record, we do actually allow it in the individual model classes anyway, so this is not a deviation from our approach. I prefer clean Jail here which matches the documentation rather than partial list to be honest.

File: app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb
Comment (ezr-ondrej) [reply app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb:35]: can you remove it then? :)

**Comment by lzap** [comment]
Removed that deleted dualstack line. Also removed the `foreman.some.host.fqdn` to `foreman.example.com` change as this breaks plugins and it is not worth doing this in this PR.

**Comment by lzap** [comment]
I fixed the tests

**Comment by lzap** [comment]
And it is FINALLY green!

**Review by ezr-ondrej — APPROVED** [review]
Thanks @lzap ! :+1:

**Pull request merged**
